### ISSUE TITLE
multi_head_attention_forward bug using torch.equal

### ISSUE THREAD

**Issue by rfernand2** [pytorch/pytorch#47979]
## 🐛 Bug

The Transformer helper function **multi_head_attention_forward** uses **torch.equal()** to compare query, key, and value vectors to each other in order to find the fastest way to compute q, k, and v.  When key, query, value contain 1 or more **nan** values, **torch.equal()** will always return **False** (even if a tensor is compared to itself).  Replacing the 3 instances of **torch.equal()** with the equivalent use of the **is** operator (`key is query`, etc.) will correctly detect the 3 different conditions.  Also, for my application, this change provided a 10% perf improvement during training.

## To Reproduce

I didn't isolate a failing simple test case for this issue, but I believe if you have a small tutorial type Transformer app that uses the Encoder/Decoder style of the **Transformer** class, and during training you pass it a bool mask for **src_key_padding_mask**, and trace with the debugger, you will see on 2nd or later Encoder layer forward() calls, the **torch.equal()** calls fail (when they should succeed, since the current call is from the encoder which passes "src, src, src," as the query, key, value params that eventually call multi_head_attention_forward).

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

Collecting environment information...
PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Enterprise
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.7.0
[pip3] torchtext==0.6.0
[pip3] torchvision==0.8.1
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              h74a9793_1
[conda] mkl                       2020.2                      256
[conda] mkl-service               2.3.0            py38hb782905_0
[conda] mkl_fft                   1.2.0            py38h45dec08_0
[conda] mkl_random                1.1.1            py38h47e9c7a_0
[conda] numpy                     1.19.2           py38hadc3359_0
[conda] numpy-base                1.19.2           py38ha3acd2a_0
[conda] pytorch                   1.7.0           py3.8_cuda102_cudnn7_0    pytorch
[conda] torchtext                 0.6.0                    pypi_0    pypi
[conda] torchvision               0.8.1                py38_cu102    pytorch

## Additional context

cc @VitalyFedyunin @ngimel @zhangguanheng66

**Comment by gchanan** [comment]
also happens in C++.

**Comment by zhangguanheng66** [comment]
Thanks @rfernand2. We can accept a PR to fix this.

**Comment by rfernand2** [comment]
I'm not set up to do pytorch source builds, testing, etc, so I would appreciate if someone who is already submitting PR's could do one for this issue.  Thanks.

**Comment by gchanan** [comment]
@zhangguanheng66 it's not clear to me if you are suggesting we replace the `equals` calls with `is` (would that be BC breaking?) or if we should do that in addition to.

**Comment by zhangguanheng66** [comment]
My understanding is that this will not change the correctness of MHA. It would be great if we can have a code snippet here to help us make a decision.

**Comment by rfernand2** [comment]
Sure, below is a copy of the code I am now using (modifies this code: https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L4128):
```
    if not use_separate_proj_weight:
        # torch.equal cannot be used for comparing tensors that contain nan values
        # this change gives a 10% perf improvement for encoder/decoder style transformer
        #if torch.equal(query, key) and torch.equal(key, value):
        if query is key and key is value:
            # self-attention
            q, k, v = linear(query, in_proj_weight, in_proj_bias).chunk(3, dim=-1)

        #elif torch.equal(key, value):
        elif key is value:
```
The first 3 and last comments can be removed (they are just to help me track any changes that I make).  Effectively, we are just replacing the 2 if statements that use **torch.equals** with 2 statements that use **is**.

This speeds up performance because the fastest path thru here is computing the q, k, v vectors with a single call to linear (the statement after the first **if**.  The **if** mistakenly returns False if the **query** or **key** vectors contain 1 or more **nan** values (even if they are the same vector).  The **if** statements were designed to detect the case where we are passing in the same vector in 2 or 3 of the query, key, value arguments, so using **is** here is a much better choice (faster to execute and not subject to **nan** containing issues).

In summary, there should be no change in functionality with the PR.  When current **if** statements mistakenly return **False**, they compute the same values for q, k, v, just using a slower approach (since they cannot assume the 2 or 3 args are the same).

**Comment by zhangguanheng66** [comment]
> Sure, below is a copy of the code I am now using (modifies this code: https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L4128):
> 
> ```
>     if not use_separate_proj_weight:
>         # torch.equal cannot be used for comparing tensors that contain nan values
>         # this change gives a 10% perf improvement for encoder/decoder style transformer
>         #if torch.equal(query, key) and torch.equal(key, value):
>         if query is key and key is value:
>             # self-attention
>             q, k, v = linear(query, in_proj_weight, in_proj_bias).chunk(3, dim=-1)
> 
>         #elif torch.equal(key, value):
>         elif key is value:
> ```
> 
> The first 3 and last comments can be removed (they are just to help me track any changes that I make). Effectively, we are just replacing the 2 if statements that use **torch.equals** with 2 statements that use **is**.
> 
> This speeds up performance because the fastest path thru here is computing the q, k, v vectors with a single call to linear (the statement after the first **if**. The **if** mistakenly returns False if the **query** or **key** vectors contain 1 or more **nan** values (even if they are the same vector). The **if** statements were designed to detect the case where we are passing in the same vector in 2 or 3 of the query, key, value arguments, so using **is** here is a much better choice (faster to execute and not subject to **nan** containing issues).
> 
> In summary, there should be no change in functionality with the PR. When current **if** statements mistakenly return **False**, they compute the same values for q, k, v, just using a slower approach (since they cannot assume the 2 or 3 args are the same).

Thanks @rfernand2. I will send out a PR and fix this.

**Comment by ngimel** [comment]
I think it should still be `or` so that fast path can be taken in both cases:
```
if (query is key or torch.equal(query, key) and (key is value or torch.equal(key, value):
```

**Comment by zhangguanheng66** [comment]
> I think it should still be `or` so that fast path can be taken in both cases:
> 
> ```
> if (query is key or torch.equal(query, key) and (key is value or torch.equal(key, value):
> ```

Will `is` cover all the cases for `torch.equal`?

**Comment by ngimel** [comment]
No, `is` returns true when tensors are literally the same object, `.equal` is when they contain equal values.

**Comment by zhangguanheng66** [comment]
> No, `is` returns true when tensors are literally the same object, `.equal` is when they contain equal values.

Thanks for the clarification.
